sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/physical/partitioning.scala

```diff
@@ -30,18 +30,32 @@ import org.apache.spark.sql.types.{DataType, IntegerType}
  *  - Intra-partition ordering of data: In this case the distribution describes guarantees made
  *    about how tuples are distributed within a single partition.
  */
-sealed trait Distribution
+sealed trait Distribution {
+  /**
+   * Given the number of partitions, return a [[Partitioning]] that can satisfy this distribution
+   * requirement. Note that the `numPartitions` is just a hint, the returned [[Partitioning]] does
+   * not have to have the same number of partitions.
+   */
+  def defaultPartitioning(numPartitions: Int): Partitioning
```

Comment:
Any better name? `satisfiedPartitioning`?

Reply:
Given that `numPartitions` only makes sense for hash and ranged partitions, is there a need to specify it as a hint in the default trait? Can we not just make it a parameter for the relevant partitioning scheme(s)?

Reply:
`numPartitions` is a parameter for the relevant partitioning scheme(s), here we need a way to create partitioning scheme for a certain distribution, and need this `numPartitions` as a hint.